calico_containers/tests/st/utils/workload.py

```diff
@@ -125,15 +128,44 @@ def assert_can_ping(self, ip, retries=0):
         command = ' '.join(args)
 
         ping = partial(self.execute, command)
-        return retry_until_success(ping,
+        return ping
+
+    def assert_can_ping(self, ip, retries=0):
+        """
+        Execute a ping from this workload to the ip. Assert than a workload
+        can ping an IP. Use retries to compensate for network uncertainty and
+        convergence.
+
+        :param ip:  The IP address (str or IPAddress) to ping.
+        :param retries: The number of retries.
+        :return: None.
+        """
+        return retry_until_success(self._get_ping_function(ip),
                                    retries=retries,
-                                   ex_class=CalledProcessError)
+                                   ex_class=CommandExecError)
 
     def assert_cant_ping(self, ip, retries=0):
-        for retry in range(retries + 1):
-            self.assert_can_ping(ip)
-            if retry >= retries:
-                raise Exception("Workload can unexpectedly ping %s" % ip)
+        """
+        Execute a ping from this workload to the ip.  Assert that the workload
+        cannot ping an IP.  Use retries to compensate for network uncertainty
+        and convergence.
+
```

Comment:
Yes I was thinking about something similar actually, but since I was moving from a ping to a no-ping scenario I wasn't convinced it was necessary ... but I guess the "no ping" could be a transient state.

I still need convincing that the assert_can_ping is immune to this problem.  For example, you have scanerio A.  Pings should work and do.  Move to scenario B ... pings should still work, so you use the assert_can_ping which succeeds.... but wait for a bit and pings stop working.  I think it is always possible to think of scenarios in both cases where stopping on the first success (be it ping or no ping) is insufficient.

IMHO, both of these should have the two sets of values.